src/Messaging/SubscriptionManager.cs

```diff
@@ -0,0 +1,84 @@
+﻿using CodeMonkeys.Core.Messaging;
+using CodeMonkeys.Messaging.Configuration;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace CodeMonkeys.Messaging
+{
+    internal sealed class SubscriptionManager
+    {
+        private readonly HashSet<Subscription> _items;
+        private readonly object _sync;
+
+        private readonly SubscriptionManagerOptions _options;
+
+        private bool _disposed;
+
+        internal SubscriptionManager(SubscriptionManagerOptions options = null)
+        {
+            _items = new HashSet<Subscription>();
+            _sync = new object();
+
+            _options = options ?? new SubscriptionManagerOptions();
+
+            Task.Run(RemoveCollectedSubscribers);
+        }
+
+        /// <inheritdoc/>
+        internal IEnumerable<ISubscriberOf<TEvent>> GetSubscribersOf<TEvent>()
+            where TEvent : class, IEvent
+        {
+            var subscriptions = _items
+                .Where(subscription => subscription.EventType.Equals(typeof(TEvent)));
+
+            return subscriptions
+                .Select(subscription => subscription.Reference.Target as ISubscriberOf<TEvent>);
+        }
+
+        /// <inheritdoc/>
+        internal void Add(Type eventType, ISubscriber subscriber)
+        {
+            lock (_sync)
+            {
+                _items.Add(
+                    Subscription.Create(
+                        eventType, 
+                        subscriber));
+            }
+        }
+
+        /// <inheritdoc/>
+        internal void Remove(Type eventType, ISubscriber subscriber)
+        {
+            lock (_sync)
+            {
+                _items.Remove(
+                    Subscription.Create(
+                        eventType, 
+                        subscriber));
+            }
+        }
+
+        private async Task RemoveCollectedSubscribers()
+        {
+            while (!_disposed)
```

Comment:
You could use a Timer aswell instead of this boolean flag and stop it on dispose. While not disposed, the timer would automatically reset.
I‘m fine with both solutions.